File: domain/src/main/java/de/x4fyr/timeline/domain/Timeline.java
package de.x4fyr.timeline.domain;

import de.x4fyr.timeline.domain.elements.ScheduledElement;

import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedList;

/**
 * Timeline aggregating scheduled elements.
 * <p>Basically this is a LinkedList, that gets sorted on add.</p>
 *
 * @param <E> target elements of this timeline.
 *
 * @author Benedikt Volkmer
 *         Created on 11/1/16.
 */
public class Timeline<E extends ScheduledElement> extends LinkedList<E> {

    private static final Comparator<ScheduledElement> COMPARATOR = Comparator.comparing(ScheduledElement::getStart);

    public Timeline() {
        super();
    }

    @SafeVarargs
    public Timeline(E... elements) {
        this();
        this.addAll(Arrays.asList(elements));
    }

    public Timeline(Collection<E> elements) {
        this();
        this.addAll(elements);
    }

    @Override
    public boolean add(E element) {
        boolean modified = super.add(element);
        super.sort(COMPARATOR);
        return modified;
    }

    @Override
    public void add(int index, E element) {
        add(element);
    }

    @Override
    public boolean addAll(Collection<? extends E> c) {
        boolean modified = super.addAll(c);
        super.sort(COMPARATOR);
        return modified;
    }

    @Override
    public boolean addAll(int index, Collection<? extends E> c) {
        boolean modified = super.addAll(size(), c);
        super.sort(COMPARATOR);
        return modified;
    }

    @Override
    public void addFirst(E element) {
        add(element);
    }

    @Override
    public void addLast(E element) {
        add(element);
    }

    @Override
    public void sort(Comparator<? super E> c) {
        super.sort(COMPARATOR);
    }

}

File: domain/src/main/kotlin/de/x4fyr/timeline/domain/Timeline.kt
package de.x4fyr.timeline.domain

import java.util.Arrays
import java.util.Comparator
import java.util.LinkedList

/**
 * Timeline aggregating scheduled elements.
 *
 * Basically this is a LinkedList, that gets sorted on add.
 * @param E target elements of this timeline.
 * @author Benedikt Volkmer
 *         Created on 11/1/16.
 */
class Timeline<E : Comparable<E>>() : LinkedList<E>() {

    private val comparator = Comparator.naturalOrder<E>()

    @SafeVarargs
    constructor(vararg elements: E) : this() {
        this.addAll(Arrays.asList(*elements))
    }

    constructor(elements: Collection<E>) : this() {
        this.addAll(elements)
    }

    /** Add element */
    override fun add(element: E): Boolean {
        val modified = super.add(element)
        sortWith(comparator)
        return modified
    }

    /** Add element
     * index will be ignored as this list is sorted
     */
    override fun add(index: Int, element: E) {
        add(element)
    }

    /** Add all elements */
    override fun addAll(elements: Collection<E>): Boolean {
        val modified = super.addAll(elements)
        sortWith(comparator)
        return modified
    }

    /** Add all elements
     * index will be ignored as this list is sorted
     */
    override fun addAll(index: Int, elements: Collection<E>): Boolean {
        val modified = super.addAll(size, elements)
        sortWith(comparator)
        return modified
    }

    /** Add element
     * It will be not necessarily be first, as this list is sorted
     */
    override fun addFirst(element: E) {
        add(element)
    }

    /** Add element
     * It will be not necessarily be last, as this list is sorted
     */
    override fun addLast(element: E) {
        add(element)
    }

    /** Replace element
     * It will no necessarily be at the same position, as this list is sorted
     */
    override fun set(index: Int, element: E): E {
        val oldElement = removeAt(index)
        add(element)
        return oldElement
    }

    /** Add element
     * It will not necessarily be first, as this list is sorted
     */
    override fun push(e: E) {
        addFirst(e)
    }

}
